anvil/distributions.py

```diff
@@ -385,38 +444,43 @@ class PhiFourAction:
 
     """
 
-    def __init__(self, m_sq, lam, geometry, use_arxiv_version=False):
-        super(PhiFourAction, self).__init__()
+    def __init__(self, geometry, parameterisation, couplings):
+        super().__init__()
+        
         self.shift = geometry.get_shift()
-        self.lam = lam
-        self.m_sq = m_sq
-        self.lattice_size = geometry.length ** 2
-        if use_arxiv_version:
-            self.version_factor = 2
-        else:
-            self.version_factor = 1
-
-    def log_density(self, phi_state: torch.Tensor) -> torch.Tensor:
-        """Perform forward pass, returning -action for stack of states. Note
-        here the minus sign since we want to return the log density of the
-        corresponding unnormalised distribution
 
-        see class Notes for details on definition of action.
-        """
+        self.__dict__.update(couplings)
+
+        if parameterisation == "standard":
+            self.nneigh_coeff = -1
+            self.quadratic_coeff = (4 + self.m_sq) / 2
+            self.quartic_coeff = self.g / 24
```

Comment:
I really dislike this. It obfuscates the initialisation of the class. The union of the different parameters is not so many, I would be strongly in favour of explicitly setting them (with a `get` method on the dictionary)

Reply:
Yeah it's horrible - I was too lazy to come up with something better at the time, but in #60 you have

```python
class PhiFourScalar:
    def __init__(
        self,
        geometry,
        ising_coefficient,
        quadratic_coefficient,
        quartic_coefficient,
    ):
        self.shift = geometry.get_shift()
        self.c_ising = ising_coefficient
        self.c2 = quadratic_coefficient
        self.c4 = quartic_coefficient

    @classmethod
    def from_standard(cls, geometry, *, m_sq, g):
        return cls(geometry, -1, (4 + m_sq) / 2, g / 24)

    @classmethod
    def from_bosetti2015(cls, geometry, *, beta, lam):
        return cls(geometry, -beta, 1 - 2 * lam, lam)

    @classmethod
    def from_albergo2019(cls, geometry, *, m_sq, lam):
        return cls(geometry, -2, 4 + m_sq, lam)
```
which I think is better than passing a parameterisation-dependent values to the constructor.

Reply:
Yeah I suppose the logic would have to go from the init to either a factory function or the production rule. I suppose I can live with the current format for now, but something like this in the not too distant future would make me sleep better 😉

Reply:
For now I've implemented the above and added logic to the 'action' instead of production rule
```python
def phi_four(geometry, parameterisation, couplings):
    constructor = getattr(PhiFourScalar, f"from_{parameterisation}")
    return constructor(geometry, **couplings)
```